Work around broken IntelSYCLConfig.cmake in Intel oneAPI 2025.x

rrsathe opened the pull request:
### Summary
Intel oneAPI 2025.x ships with a faulty `IntelSYCLConfig.cmake` file that causes the CMake configuration to fail when `find_package(IntelSYCL)` is invoked. This failure is due to an incorrect use of `SYCL_FEATURE_TEST_EXTRACT()`.

As a result, building `llama.cpp` with SYCL on Windows using the latest Intel compilers is not feasible, despite the compilers (`icx/icpx`) fully supporting SYCL.

---

### Solution
This pull request introduces a workaround in `ggml/src/ggml-sycl/CMakeLists.txt` that includes the following changes:

- Bypasses `find_package(IntelSYCL)` by default to avoid issues associated with the faulty Intel CMake package.
- Enables SYCL manually using the `-fsycl` flags, which are supported by both `icx` and the open-source `clang++`.
- Introduces an option called `GGML_SYCL_USE_FIND_PACKAGE` (default is OFF) that allows users to opt-in for testing older or fixed Intel releases.
- Implements defensive checks and provides more explicit diagnostic messages.
- Removes unsafe compiler reassignments during the configuration process.

No system files are modified, and all existing SYCL targets (Intel, NVIDIA, AMD) continue to be supported.

---

### Behavior
- **Default Method:** SYCL is enabled manually using `-fsycl`.
- **Opt-in Method:** To use `find_package(IntelSYCL)`, set `-DGGML_SYCL_USE_FIND_PACKAGE=ON`.
- **Scope:** This change is limited to CMake configuration only and does not impact runtime or backend logic.

---

### Testing
Verification has been conducted on:
- Windows 11
- Intel oneAPI 2025.3.0
- CMake with Ninja

Results:
- CMake configuration completes successfully.
- The SYCL compiler (`icx`) is detected correctly.
- The build proceeds as expected using the fallback SYCL path.

---

### AI Usage Disclosure
AI assistance was used in drafting and refining the CMake workaround and the pull request description. All changes were manually reviewed, tested on Windows with Intel oneAPI 2025.3.0, and validated by the author.

rrsathe committed:
cmake: work around broken IntelSYCLConfig.cmake in oneAPI 2025.x

NeoZhangJianyu reviewed (changes requested):
Which oneAPI version will trigger the issue?
I remember an old oneAPI has this issue and it has been fixed in another release.

I can't find the issue when build llama.cpp with the latest version (2025.3.1) of oneAPI in windows.

rrsathe commented:
> Which oneAPI version will trigger the issue? I remember an old oneAPI has this issue and it has been fixed in another release.
> 
> I can't find the issue when build llama.cpp with the latest version (2025.3.1) of oneAPI in windows.

Thank you for checking.

I was able to reproduce the failure on Windows with oneAPI version 2025.3.0 (using icx/icpx) when the `find_package(IntelSYCL)` command was enabled. In that configuration, CMake encounters an error inside the `IntelSYCLConfig.cmake` file due to an invalid invocation of `SYCL_FEATURE_TEST_EXTRACT()`.

However, when I applied the workaround by skipping the `find_package(IntelSYCL)`, the same environment configured and built successfully, as shown in the attached log (with IntelLLVM 2025.3.0, CMake, and Ninja).

It’s possible that Intel fixed this issue silently in version 2025.3.1, which could explain why you are no longer experiencing the problem. The change made here is intentionally defensive and can be reversed by opting in (using `GGML_SYCL_USE_FIND_PACKAGE`), in order to prevent hard failures for users on affected versions of 2025.x releases.

NeoZhangJianyu commented on ggml/src/ggml-sycl/CMakeLists.txt:
```diff
@@ -30,17 +32,32 @@ target_sources(ggml-sycl PRIVATE ${GGML_HEADERS_SYCL} ${GGML_SOURCES_SYCL})
 if (WIN32)
     # To generate a Visual Studio solution, using Intel C++ Compiler for ggml-sycl is mandatory
     if( ${CMAKE_GENERATOR} MATCHES "Visual Studio" AND NOT (${CMAKE_GENERATOR_TOOLSET} MATCHES "Intel C"))
+        if (NOT CMAKE_CXX_COMPILER_ID MATCHES "IntelLLVM")
```
In windows, the C++ compiler is icx too.
This part shouldn't be changed.

NeoZhangJianyu reviewed (changes requested):
find_package(IntelSYCL) wrong is a very critical issue.
That means oneAPI package can't be used to build code possiblly.
The fault maybe impact the build event if you set the compiler parameter manually.

The workaround solution in this PR won't work well for other cases in the future.

Maybe return error and push user to use good oneAPI release is better method.
It will avoid more potential issues in the bad oneAPI release.

rrsathe committed:
[AI] sycl: auto-detect and skip incompatible IntelSYCL package

Automatically detect compiler versions with incompatible IntelSYCL
CMake configuration files and fall back to manual SYCL flags instead
of requiring users to set options manually.

Fixes build failures with oneAPI 2025.x where IntelSYCLConfig.cmake
has SYCL_FEATURE_TEST_EXTRACT invocation errors.

rrsathe committed:
Merge branch 'ggml-org:master' into fix/oneapi-2025-sycl-cmake

rrsathe committed:
Merge branch 'fix/oneapi-2025-sycl-cmake' of https://github.com/rrsathe/llama.cpp into fix/oneapi-2025-sycl-cmake

rrsathe committed:
refactor: improve SYCL provider handling and error messages in CMake configuration

rrsathe committed:
refactor: enhance SYCL provider validation and error handling in CMake configuration

rrsathe commented on ggml/src/ggml-sycl/CMakeLists.txt:
```diff
@@ -1,32 +1,114 @@
+        set(SYCL_PKG_TEST_DIR "${CMAKE_CURRENT_BINARY_DIR}/sycl_pkg_check")
+        file(MAKE_DIRECTORY "${SYCL_PKG_TEST_DIR}")
+        file(WRITE "${SYCL_PKG_TEST_DIR}/CMakeLists.txt"
+            "cmake_minimum_required(VERSION 3.12)\n"
```
Resolved.

NeoZhangJianyu reviewed (commented):
Thank you for your hard work to fix this issue!

The solution is a little complex and hard to test and maintain, because so many branches in the code.

Here is my suggestion.
The root cause is the sentence `find_package(IntelSYCL)` would be fault and break the build with bad oneAPI.

The function in this PR could detect the `find_package(IntelSYCL)` is passed or not.

Following is the original code of this part.
We just need to enhance `find_package(IntelSYCL)` by existed solution in this PR to set `IntelSYCL_FOUND` to false when oneAPI is bad, instead of break the build.

So draft a new function (like detect_and_find_package(IntelSYCL)) to replace  `find_package(IntelSYCL)` only. Other code won't be impacted.

```
find_package(IntelSYCL) ->detect_and_find_package(IntelSYCL)
if (IntelSYCL_FOUND)
    # Use oneAPI CMake when possible
    target_link_libraries(ggml-sycl PRIVATE IntelSYCL::SYCL_CXX)
else()
    # Fallback to the simplest way of enabling SYCL when using intel/llvm nightly for instance
    target_compile_options(ggml-sycl PRIVATE "-fsycl")
    target_link_options(ggml-sycl PRIVATE "-fsycl")
endif()
```

rrsathe committed:
ggml-sycl: wrap find_package(IntelSYCL) to prevent build crashes

rrsathe committed:
Merge branch 'ggml-org:master' into fix/oneapi-2025-sycl-cmake

NeoZhangJianyu reviewed (approved):
It's great work!

I have tested on Linux.
After I make the IntelSYCL is wrong, the cmake will build correctly.

Thank you!
